docs/config_system.md

```diff
@@ -0,0 +1,270 @@
+             "timer_period": 100,
+             "queue_size": 40
+        },
+        "NXP": {
+             "queue_size": 20,
+             "buffer_size": 128
+        }
+    }
+}
+```
+
+In this JSON file:
+
+- `name` is the name of the library. **This is a required field.**
+- `config` defines the configuration parameters of the library, as explained [here](#defining-configuration-parameters).
+- `macros` is a list of extra macros that will be defined when compiling a project that includes this library. A macro can be defined without a value (like `MYMOD_MACRO1` above) or with a value (like `MYMOD_MACRO2` above).
+- `target_overrides` is a dictionary with target-specific values for the configuration parameters.
+
+`target_overrides` is used to override the values of the parameters depending on the current compilation target. The keys in `target_overrides` are matched against toolchain *labels* (a description of mbed targets can be found [here](mbed_targets.md)). If a key inside `target_overrides` matches one of the target labels, the parameter values are changed according to the value of the key. In the example above:
+
+- `config` is always processed first, independent of the target. `config` might define values for some of the parameters. In this case, `buffer_size` will be set to 1024, `queue_size` will be set to 10 and `timer_period` will not have a value.
+- if the library is compiled for the `K64F` target, `timer_period` will be set to 100 and `queue_size` will be set to 40, since they are overridden by the `K64F` key in `target_overrides`. `buffer_size` will be set to 1024, as defined in `config`.
+- assuming that `NXP` is a label defined by **all** NXP based targets, if the library is compiled for **any** `NXP` target (like `LPC1768` or `LPC11U24`), `buffer_size` will be set to 128 and `queue_size` will be set to 20, while `timer_period` will not have a value (since it doesn't get one neither in `config`, nor in the `NXP` override).
+- the keys in `target_overrides` are processed in order: if a hypothetical target defines both `K64F` and `NXP` as labels, `timer_period` will be set to 100, `queue_size` will be set to 20 and `buffer_size` will be set to 128.
+- if the library is compiled for a target that doesn't have `K64F` or `NXP` as labels, the values of the parameters will be the ones set in `config`.
+
+Except `name`, all the above keys in the JSON file are optional, but if `target_overrides` is defined, `config` must also be defined.
+
+As explained [here](#defining-configuration-parameters), the parameters have an implicit `mylib.` prefix. Outside `mylib`, `buffer_size` is accessible using the name `mylib.buffer_size`. An application will be able to override the value of this parameter, as described in [this section](#configuration-data-in-applications).
+
+ If the source tree has code for more than one library, each library needs its own `mbed_lib.json` file in its root folder.
+
+# Configuration data in targets
```

Comment:
Targets already provide base configuration for generic macros, labels, features and device peripheral configuration (via device_has).

If this is the case then we need to turn ALL targets in targets.json into configuration data and I don't think that this is sensible or needed. I'd rather expect that the targets.json already implement common data which is well documented.

-1